Gateway docs (#1745)
solo-io/gloo-mesh: docs/content/guides/gateway/concepts.md

---
title: Gateway Concepts
menuTitle: Concepts
description: A walkthrough of the Gloo Mesh Gateway configuration resources.
weight: 30
---

{{% notice note %}} Basic Gateway features are available to anyone with a Gloo Mesh Enterprise license. Some advanced features require a Gloo Mesh Gateway license. {{% /notice %}}

## Gloo Mesh Gateway

> Gloo Mesh Gateway is an abstraction built on top of Istio ingress gateway, and Envoy proxy. It can be used as a way to simplfy the configuration of ingress traffic rules, while retaining seamless integration with your service mesh. It also adds powerful mutli-cluster, multi-mesh capabilities for scalability.

Reviewer: ```suggestion
Gloo Mesh Gateway is an abstraction built on top of [Istio's ingress gateway model](https://istio.io/latest/docs/tasks/traffic-management/ingress/ingress-control/) and the underlying Envoy proxies. It can be used as a way to simplify the configuration of ingress traffic rules while retaining seamless integration with your service mesh. It also adds powerful mutli-cluster, multi-mesh capabilities for scalability.
```

## Example Architectures

There are many ways to set up ingress gateways into your service mesh, particularly when you scale up to multiple clusters and multiple meshes. Below we have a few examples of what this could look like.

### Example 1: Single cluster, Single Gateway

// TODO: Insert architecture diagram.

In this setup, we have a single ingress gateway routing traffic to a single cluster.

> ### Example 2: Mutli cluster, Single Gateway

Reviewer: ```suggestion
### Example 2: Multicluster, Single Gateway
```

// TODO: Insert architecture diagram.

In this setup, our application has grown and now spans two meshes deployed across two clusters. We still have a single ingress gateway configured, but notably it is aware of services running in both meshes and both clusters. Having the application split like this can help with separation of concerns when multiple application teams are involved. It also offers better resilience and high availability, should an application in one cluster fail.

> ### Example 3: Mutli cluster, Multi Gateway

Reviewer: ```suggestion
### Example 3: Multicluster, Multi Gateway
```

## Gloo Mesh Gateway Custom Resources

In order to make all of the above scenarios easier to configure, we have introduced three new Custom Resources (CRs):

- VirtualGateway
- VirtualHost
- RouteTable

Each of these resources will live in the Management Cluster, although they may be used to configure ingress resources in any cluster managed by Gloo Mesh. These resources are installed as part of Gloo Mesh Enterprise, and no additional settings are required to start using Gateway features.

### VirtualGateway

> A VirtualGateway resource will define the listeners, which protocols and ports to listen on. It will also define which filters are attached, by allowing users to specify a FilterChain (or multiple FilterChains). It attaches to existing Ingress Gateway workloads via a workload selector, and specifies one or more VirtualHost configurations to configure routing rules. A single VirtualGateway can apply to multiple ingress deployments across meshes and clusters contained within a VirtualMesh. Attached Gateways will be listed in the status of the VirtualGateway. In order to perform cross-mesh routing, the Gateway Mesh and Destination Mesh must be contained in a single VirtualMesh, with federation enabled.

Reviewer: ```suggestion
A VirtualGateway resource will define the [Envoy listeners](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/listeners), i.e. the protocols and ports to listen on. It will also define which Envoy filters are attached, by allowing users to specify a [FilterChain](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/listeners/network_filter_chain) (or multiple FilterChains). It attaches to existing ingress gateway workloads via a workload selector, and specifies one or more VirtualHost configurations to configure routing rules. A single VirtualGateway can apply to multiple ingress deployments across meshes and clusters contained within a VirtualMesh. Attached Gateways will be listed in the status of the VirtualGateway. In order to perform cross-mesh routing, the Gateway Mesh and Destination Mesh _must be contained in a single VirtualMesh with federation enabled_.
```

### VirtualHost

> Virtual Hosts are selected by a Virtual Gateway. A single VirtualHost can be selected by multiple VirtualGateways. VirtualHosts are responsible for configuring top level settings, such as domains. In addition, a VirtualHost can set route options that apply to all child routes, which will be inherited as a default unless explicitly overridden at the route level. A Virtual Host contains a list of Routes, which can contain various matchers and actions. 

Reviewer: Can we do a find+replace for all occurrences of "Virtual Host" and "Virtual Hosts" and replace with "VirtualHost" and "VirtualHosts"? Same for "Virtual Gateway(s)" and "Route Table(s)"

### RouteTable

> A Route Table is effectively a list of Routes. It exists only to be delegated to from VirtualHosts, or other RouteTables. This allows users to separate configuration and ownership across the organization. For example an app-level team may want to configure all of the low-level endpoints that are sent to their service, but may not have control over which domains they can serve traffic on, or what the auth policies are. Routes configured in RouteTables will by default inherit any options configured by their parent delegating resource (RouteTable or VirtualHost), for example, timeout or retry settings.

Reviewer: ```suggestion
A Route Table is effectively a list of Routes. It exists only to be delegated to from VirtualHosts, or other RouteTables. This allows users to separate configuration and ownership across the organization. For example an app-level team may want to configure all of the low-level endpoints that are sent to their service, but may not have control over which domains they can serve traffic on, or what the authorization policies are. Routes configured in RouteTables will by default inherit any options configured by their parent delegating resource (RouteTable or VirtualHost), for example, timeout or retry settings.
```

### Routes

Routes can be configured on a VirtualHost, and/or a RouteTable. Routes consist of options, matchers, and an action. Options are a way to specify route policies in-line, for example a retry policy or a timeout. Matchers are how the routing logic determines which route is selected. It could match on a path, http method, or header, for example. Routes can have one of four actions:

#### Route Action

Used to send traffic directly to one of three destination types:

> - A kubernetes service, the most commonly used option. Notably, this is a service in a specific _cluster_.

Reviewer: Make sure to capitalize all occurrences of "kubernetes"

```suggestion
- A Kubernetes service, the most commonly used option. Notably, this is a service in a specific _cluster_.
```

---
title: Gateway Concepts Overview
menuTitle: Concepts
description: A walkthrough of the Gloo Mesh Gateway configuration resources.
weight: 10
---

{{% notice note %}} Basic Gateway features are available to anyone with a Gloo Mesh Enterprise license. Some advanced features require a Gloo Mesh Gateway license. {{% /notice %}}

## Gloo Mesh Gateway

> Gloo Mesh Gateway is an abstraction built on top of [Istio's ingress gateway model](https://istio.io/latest/docs/tasks/traffic-management/ingress/ingress-control/) and the underlying Envoy proxies. It can be used as a way to simplify the configuration of ingress traffic rules while retaining seamless integration with your service mesh. It also adds powerful mutli-cluster, multi-mesh capabilities for scalability.

Reviewer 2: ```suggestion
Gloo Mesh Gateway is an abstraction built on top of [Istio's ingress gateway model](https://istio.io/latest/docs/tasks/traffic-management/ingress/ingress-control/). It can be used as a way to simplify the configuration of ingress traffic rules while retaining seamless integration with your service mesh. It also adds powerful mutli-cluster, multi-mesh capabilities to your gateway.
```

Reviewer 2: Is multi-mesh here, multi istio? If so should we be explicit about that?

Author: For now, yes - although I don't think we've definitively ruled out supporting other meshes for this in the long term.

## Example Architectures

There are many ways to set up ingress gateways into your service mesh, particularly when you scale up to multiple clusters and multiple meshes. Below we have a few examples of what this could look like.

### Example 1: Single cluster, Single Gateway

![Single Cluster Single Gateway]({{% versioned_link_path fromRoot="/img/gateway/gateway-single-cluster.png" %}})

In this setup, we have a single ingress gateway routing traffic to a single cluster.

### Example 2: Single cluster, Multi Gateway

![Single cluster, Multi Gateway]({{% versioned_link_path fromRoot="/img/gateway/gateway-single-cluster-multi-gateway.png" %}})

> In this setup, we have a single cluster deployed with multiple gateways. The gateways can have overlap in which services they route traffic to, if desired.

Reviewer 2: ```suggestion
In this setup, we have a single cluster deployed with multiple gateways. The services which these gateways route traffic to can overlap, if desired.
```

## Gloo Mesh Gateway Custom Resources

In order to make all of the above scenarios easier to configure, we have introduced three new Custom Resources (CRs):

- VirtualGateway
- VirtualHost
- RouteTable

Each of these resources will live in the Management Cluster, although they may be used to configure ingress resources in any cluster managed by Gloo Mesh. These resources are installed as part of Gloo Mesh Enterprise, and no additional settings are required to start using Gateway features.

### VirtualGateway

> A VirtualGateway resource will define the [Envoy listeners](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/listeners), i.e. the protocols and ports to listen on. It will also define which Envoy filters are attached, by allowing users to specify a [FilterChain](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/listeners/network_filter_chain) (or multiple FilterChains). It attaches to existing ingress gateway workloads via a workload selector, and specifies one or more VirtualHost configurations to configure routing rules. A single VirtualGateway can apply to multiple ingress deployments across meshes and clusters contained within a VirtualMesh. Attached Gateways will be listed in the status of the VirtualGateway. In order to perform cross-mesh routing, the Gateway Mesh and Destination Mesh _must be contained in a single VirtualMesh with federation enabled_.

Reviewer 2: ```suggestion
A VirtualGateway resource will define the [Envoy listeners](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/listeners), i.e. the protocols and ports to listen on. It will also define which Envoy filters are attached, by allowing users to specify a [FilterChain](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/listeners/network_filter_chain) (or multiple FilterChains). It attaches to existing ingress gateway workloads via a workload selector, and specifies one or more VirtualHost configurations to configure routing rules. A single VirtualGateway can apply to multiple ingress gateway deployments across all meshes and clusters contained within a VirtualMesh. Attached Gateways will be listed in the status of the VirtualGateway. In order to perform cross-mesh routing, the Gateway Mesh and Destination Mesh _must be contained in a single VirtualMesh with federation enabled_.
```

This description is probably a bit low level, the user doesn't care about listeners or filter chains. The first bit reads more like internal documentation

### VirtualHost

> VirtualHosts are selected by a VirtualGateway. A single VirtualHost can be selected by multiple VirtualGateways. VirtualHosts are responsible for configuring top level settings, such as domains. In addition, a VirtualHost can set route options that apply to all child routes, which will be inherited as a default unless explicitly overridden at the route level. A VirtualHost contains a list of Routes, which can contain various matchers and actions. 

Reviewer 2: What do you mean by "top level settings"?

Author: I meant settings that can only be configured once for a VirtualHost, but will affect all child routes. Things like domain for example. Which is slightly different from setting route options on the VirtualHost, which act more like a default for child routes, but still lets the child routes overwrite them on a case by case basis if desired. Really "VirtualHost" level settings, but I'm open to suggestions on how we phrase that without using the word VirtualHost, as that's what we're trying to introduce here :)
